Use inline Vector{128|256}.Create for constants

gfoidl opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/44115

Used `Vector(128|256)<\w+> \w+ =` as regex-search, seems that these three places are the onley one left.
The SSE-helper, that are mentioned in the issue, got re-written by @GrabYourPitchforks some time ago, and there patterns for the new codegen is already used.

Codegen was validated (manually) to see that it's an improvement.

It's not clear to me whether the touched codes are built for .NET Core 3.1 or not. So I added conditional compilation, as for .NET Core 3.1 the codegen is quite worse when Vector{128|256}.Create is used (that's why we initially had the different patterns before the JIT started to emit constants and read them from the data-segment).

/cc: @tannergooding

gfoidl committed:
Used VectorXYZ.Create for constants in Base64

gfoidl committed:
Used VectorXYZ.Create for constants in BitArray

stephentoub commented on src/libraries/System.Collections/src/System/Collections/BitArray.cs:
```diff
@@ -120,9 +120,11 @@ public BitArray(byte[] bytes)
             _version = 0;
         }
 
+#if !NET6_0_OR_GREATER
```
Why are these ifdefs needed... isn't this file only compiled for netcoreappcurrent?

gfoidl replied:
> isn't this file only compiled for netcoreappcurrent?

This is the point I don't know (alluded in the description).
But if so, I'll just remove them.

My point is just to make sure that .NET Core 3.1 doesn't regress, as it's a LTS through Dec, 2022.

stephentoub replied:
They're not needed.  This file is only included in System.Collections.csproj, which only builds for netcoreappcurrent:
https://github.com/dotnet/runtime/blob/bde0322a9e11fba40a7a65c6d38200d6514efc72/src/libraries/System.Collections/src/System.Collections.csproj#L5

Base64Decoder/Encoder.cs are only included in System.Memory, which only builds for netcoreappcurrent:
https://github.com/dotnet/runtime/blob/bde0322a9e11fba40a7a65c6d38200d6514efc72/src/libraries/System.Memory/src/System.Memory.csproj#L6

gfoidl replied:
Thanks! 👍🏻

stephentoub commented on src/libraries/System.Memory/src/System/Buffers/Text/Base64Decoder.cs:
```diff
@@ -362,14 +362,64 @@ private static unsafe void Avx2Decode(ref byte* srcBytes, ref byte* destBytes, b
             // See SSSE3-version below for an explanation of how the code works.
 
             // The JIT won't hoist these "constants", so help it
+#if NET6_0_OR_GREATER
```
Same question

stephentoub commented on src/libraries/System.Memory/src/System/Buffers/Text/Base64Encoder.cs:
```diff
@@ -241,14 +241,36 @@ private static unsafe void Avx2Encode(ref byte* srcBytes, ref byte* destBytes, b
             // l k j i h g f e d c b a 0 0 0 0
 
             // The JIT won't hoist these "constants", so help it
+#if NET6_0_OR_GREATER
```
Ditto.  I don't think any of these ifdefs are needed.

gfoidl committed:
Remove conditional compilation

It's only built for NetCoreAppCurrent, so no need to special case older runtimes.
